File: task2/.php
<?php

function normalize_email_php(string $email): string {
    $parts = explode('@', $email); // Разбиваем email на имя пользователя и домен по символу @
    if (count($parts) != 2) {
        return ""; // Если не 2 части (нет @), возвращаем пустую строку (некорректный формат)
    }
    $username = $parts[0]; // имя пользователя
    $domain = $parts[1]; // домен

    $normalized_username = ""; // Инициализируем строку для нормализованного имени пользователя
    for ($i = 0; $i < strlen($username); $i++) { // Перебираем символы в имени пользователя
        $char = $username[$i]; // // Перебираем символы в имени пользователя
        if ($char == '.') {
            continue;
        } else if ($char == '+' || $char == '*') {
            break;
        } else {
            $normalized_username .= $char; // Добавляем символ в нормализованное имя пользователя
        }
    }
    return $normalized_username . "@" . $domain; // Возвращаем нормализованный email
}

function count_unique_emails_php(array $emails): int {
    $unique_emails = []; // Инициализируем пустой массив для нормализованных email-адресов
    foreach ($emails as $email) { // Перебираем email-адреса в массиве
        $unique_emails[] = normalize_email_php($email); // Нормализуем email и добавляем в массив
    }
    return count(array_unique($unique_emails));// использую array для удаления дубликатов
}

$emails1 = ["mar.pha+science@corp.nstu.ru", "marpha+scie.nce@corp.nstu.ru", "marph.a+s.c.i.e.n.c.e+@corp.nstu.ru"];
echo "Вывод 1 Output: " . count_unique_emails_php($emails1) . "\n";

$emails2 = ["mar.pha+science@co.rp.nstu.ru", "marpha+scie.nce@corp.nstu.ru", "marph.a+s.c.i.e.n.c.e+@corp.nstu.ru"];
echo "Вывод 2 Output: " . count_unique_emails_php($emails2) . "\n";

?>

File: task2/.swift
import Foundation 

func normalizeEmailSwift(_ email: String) -> String { // Функция нормализации email
    guard let atPos = email.firstIndex(of: "@") else { // Находим индекс символа @
        return "" // Если @ нет, возвращаем пустую строку
    }
    let username = String(email[..<atPos]) // Имя пользователя (до @)
    let domain = String(email[atPos...].dropFirst()) // Домен (после @)

    var normalizedUsername = "" // Строка для нормализованного имени
    for char in username { // Перебираем символы имени
        if char == "." { 
            continue
        } else if char == "+" || char == "*" { 
            break
        } else {
            normalizedUsername.append(char) // Добавляем остальные символы
        }
    }
    return normalizedUsername + "@" + domain // Возвращаем нормализованный email
}

func countUniqueEmailsSwift(_ emails: [String]) -> Int { // Функция подсчета уникальных email
    var uniqueEmails = Set<String>() // Множество для уникальных email
    for email in emails { // Перебираем список email
        uniqueEmails.insert(normalizeEmailSwift(email)) // Нормализуем и добавляем в множество
    }
    return uniqueEmails.count // Возвращаем количество уникальных email
}

let emails1 = ["mar.pha+science@corp.nstu.ru", "marpha+scie.nce@corp.nstu.ru", "marph.a+s.c.i.e.n.c.e+@corp.nstu.ru"] // Пример 1
print("Example 1 Output: \(countUniqueEmailsSwift(emails1))") 

let emails2 = ["mar.pha+science@co.rp.nstu.ru", "marpha+scie.nce@corp.nstu.ru", "marph.a+s.c.i.e.n.c.e+@corp.nstu.ru"] // Пример 2
print("Example 2 Output: \(countUniqueEmailsSwift(emails2))")
